File: partitionToKEqualSumLC698/Solution.py

# Accepted Python version by Leetcode
# https://leetcode.com/problems/partition-to-k-equal-sum-subsets/submissions/850211390/
# https://leetcode.com/problems/partition-to-k-equal-sum-subsets/submissions/851413176/

# Based on Neetcode's code plus optimization from comments by Rg Prasad
# https://www.youtube.com/watch?v=mBk4I0X46oI&list=PLot-Xpze53lf5C3HSjCnyFghlW0G1HHXo&index=16

# this version has same number value optimization.
# For example, if input is { 1, 1, 1, 1, 2, 2, 2, 2 }, k=4, and if a previous "1" or "2" is not 
# used for bucket sum, then next "1" or "2" can't be used either ("not used[i - 1] and nums[i] == nums[i - 1]")
# The reason is when checking a previous "1" or "2", it already checked the cases for including the next "1" or "2",
# so if the previous number is NOT selected for target sum, then it means including the next "1" or "2" won't work either     

# this version was accepted by Leetcode, but seems not very fast:
#Runtime 707 ms  Beats 49.19% Memory 13.9 MB Beats 93.65%

#Runtime 408 ms Beats 61.12% Memory 13.9 MB Beats 74%

from typing import List


class Solution:
    def canPartitionKSubsets(self, nums: List[int], k: int) -> bool:
        
        n = len(nums)
        nums.sort(reverse=True)
        total = sum(nums)
        if total % k != 0:
            return False
        
        target = int(total / k)
        used = [False] * n
        
        def dfs(index, total, k): 
            
            if k == 0:
                return True
            
            if total == 0:
                return dfs(0, target, k - 1)
            
            for i in range(index, n):
                
                # Rg Prasad added below optimization to avoid duplicate calculations on i
                # if i-1 has already checked the same case (target, inputs)
                if i > 0 and not used[i - 1] and nums[i] == nums[i - 1]:
                    continue
                    
                if used[i] or total - nums[i] < 0:
                    continue
                
                used[i] = True
                if dfs(i + 1, total - nums[i], k):
                    return True
                used[i] = False
            return False
        
        return dfs(0, target, k)
    
    
s = Solution()
print(" result =", s.canPartitionKSubsets([10, 10, 7, 8, 10, 4, 9, 7, 9, 10, 4, 6, 7, 1, 8, 5], 5))    

File: partitionToKEqualSumLC698/Solution.java
package partitionToKEqualSumLC698;

import java.util.*;

// https://leetcode.com/problems/partition-to-k-equal-sum-subsets/description/

// First working version by using binary decision tree instead of non-binary dfs tree
// used by Neetcode and others:
// https://www.youtube.com/watch?v=mBk4I0X46oI&list=PLot-Xpze53lf5C3HSjCnyFghlW0G1HHXo&index=16 (Python)
// https://www.youtube.com/watch?v=8t8TeyRJDvk (Java code: https://github.com/Sunchit/Coding-Decoded/blob/master/September2021/PartitiontoKEqualSumSubsets.java)

// This version is based on binary decision tree( pick a number / not pick a number). It's different from Neetcode's version, which essentially a compressed decision tree.  

// This version was working, but performance was bad, got following below TLE error:
// Time Limit Exceeded
// Last Executed Input
// 151 / 162 testcases passed
// nums =
// [3,9,4,5,8,8,7,9,3,6,2,10,10,4,10,2]
// k =
// 10

public class Solution {

    public static void main(String[] args) {
        Solution solution = new Solution();
        // boolean res = solution.canPartitionKSubsets(new int[] {4,3,2,3,5,2,1}, 4);
        // boolean res = solution.canPartitionKSubsets(new int[] {2,2,2,2,3,4,5}, 4);
        boolean res = solution.canPartitionKSubsets(new int[] { 1, 1, 1, 1, 2, 2, 2, 2 }, 4);

        System.out.println(res);
    }

    int[] nums;
    int SUM;
    List<Integer> usedNums = new ArrayList<>();

    public boolean canPartitionKSubsets(int[] nums, int k) {
        if (nums == null || nums.length == 0) {
            return false;
        }

        SUM = Arrays.stream(nums).sum();

        if (SUM % k != 0) {
            return false;
        }
        SUM /= k;

        Arrays.sort(nums);
        reverse(nums);
        if ((nums[0]) > SUM) {
            return false;
        }

        this.nums = nums;

        return partition(k, 0, 0);
    }

    private boolean partition(int buckets, int numPos, int bucketSum) {
        if (buckets == 0) {
            return true;
        }

        if (bucketSum == SUM) {
            boolean res = partition(buckets - 1, 0, 0);
            return res;
        }

        while (usedNums.contains(numPos)) {
            ++numPos;
        }

        if (numPos >= nums.length) {
            return false;
        }

        // if (numPos > 0 && nums[numPos] == nums[numPos - 1] &&
        // !usedNums.contains(numPos - 1)) {
        // return false;
        // }

        if (bucketSum + nums[numPos] <= SUM) {

            usedNums.add(numPos);
            if (partition(buckets, numPos + 1, bucketSum + nums[numPos])) {
                return true;
            }
            ;
            usedNums.remove(usedNums.size() - 1);
        }

        if (partition(buckets, numPos + 1, bucketSum)) {
            return true;
        }

        return false;
    }

    private void reverse(int[] nums) {
        for (int i = 0; i < nums.length / 2; ++i) {
            int tmp = nums[i];
            nums[i] = nums[nums.length - 1 - i];
            nums[nums.length - 1 - i] = tmp;
        }
    }
}
